src/Land/Model/SurfaceFlow.jl

```diff
@@ -0,0 +1,248 @@
+    aux,
+    t,
+) end
+
+"""
+    VolumeAdvection <: TendencyDef{Flux{FirstOrder}}
+
+A first order flux type for the overland flow model.
+"""
+struct VolumeAdvection <: TendencyDef{Flux{FirstOrder}} end
+
+
+"""
+    flux(::SurfaceWaterArea, ::VolumeAdvection, land::LandModel, args,)
+
+A first order flux method for the OverlandFlow model, adding in advection of water volume due to flowing water.
+"""
+function flux(::SurfaceWaterArea, ::VolumeAdvection, land::LandModel, args)
+    @unpack state, aux = args
+    x = aux.x
+    y = aux.y
+    width = land.surface.width(x, y)
+    area = max(eltype(state)(0.0), state.surface.area)
+    height = area / width
+    v = calculate_velocity(land.surface, x, y, height)
+    Q = area * v
+    return Q
+end
+
+# Boundary Conditions
+
+# General case - to be used with bc::NoBC
+function surface_boundary_flux!(
+    nf,
+    bc::Land.AbstractBoundaryConditions,
+    m::Union{NoSurfaceFlowModel, OverlandFlowModel},
+    land::LandModel,
+    _...,
+) end
+
+function surface_boundary_state!(
+    nf,
+    bc::Land.AbstractBoundaryConditions,
+    m::Union{NoSurfaceFlowModel, OverlandFlowModel},
+    land::LandModel,
+    _...,
+) end
+
+"""
+    surface_boundary_flux!(
+        nf,
+        bc::Land.Dirichlet,
+        model::OverlandFlowModel,
+        land::LandModel,
+        _...,
+    )
+
+The surface boundary flux function for the OverlandFlow model, which
+does nothing if Dirichlet conditions are chosen in order to not
+overconstrain the solution.
+"""
+function surface_boundary_flux!(
+    nf,
+    bc::Land.Dirichlet,
+    model::OverlandFlowModel,
+    land::LandModel,
+    _...,
+) end
+
+"""
+    surface_boundary_state!(
+        nf,
+        bc::Land.Dirichlet,
+        model::OverlandFlowModel,
+        land::LandModel,
+        state⁺::Vars,
+        aux⁺::Vars,
+        nM,
+        state⁻::Vars,
+        aux⁻::Vars,
+        t,
+        _...,
+    )
+
+The surface boundary state function for the OverlandFlow model when
+Dirichlet conditions are chosen. This should be equivalent to 
+outflow boundary conditions, also referred to as gradient outlet
+conditions.
+"""
+function surface_boundary_state!(
+    nf,
+    bc::Land.Dirichlet,
+    model::OverlandFlowModel,
+    land::LandModel,
+    state⁺::Vars,
+    aux⁺::Vars,
+    nM,
+    state⁻::Vars,
+    aux⁻::Vars,
+    t,
+    _...,
+)
+    bc_function = bc.state_bc
+    state⁺.surface.area = bc_function(aux⁻, t)
+end
+
+"""
+    Precip <: TendencyDef{Source}
+
+A source term for overland flow where a prescribed net precipitation
+drives the overland flow.
+"""
+struct Precip{FT, F} <: TendencyDef{Source}
+    precip::F
+
+    function Precip{FT}(precip::F) where {FT, F}
+        new{FT, F}(precip)
+    end
+end
+
+function (p::Precip{FT})(x, y, t) where {FT}
+    FT(p.precip(x, y, t))
+end
+
+
+prognostic_vars(::Precip) = (SurfaceWaterArea(),)
+
+
+function precompute(source_type::Precip, land::LandModel, args, tt::Source)
+    nothing
+end
```

Comment:
`precompute` should always return a `NamedTuple`

Reply:
Also, there's a fallback, so we generally don't need to define this unless it returns something (unless you need a fallback with a different number of arguments). I think we could probably define a generic fallback for 4 arguments like this also (i.e., make this definition generic and move to `BalanceLaws/`)

Reply:
ah ok. Ill change that! also, look like there is no fallback method for a `source_type` that does nothing? i.e. if not all sources need to precompute something.

Reply:
(I see a fallback for `precompute(bl, args, source)` only)

Reply:
Yeah, fine to leave it here for now, but instead return a `NamedTuple`. We can make it generic and move to BalanceLaws layer

Reply:
sorry, some crosstalk while the page was not reloaded :) thanks for the feedback!